### DISCUSSION TITLE
How are the unique ID's of objects generated in the XML?

### DISCUSSION THREAD

**Discussion by securitymonster** [jgraph/drawio#4333] [General]
I have documents with many pages and use properties and place-holders on these pages, but they are the same for each page. Currently properties are unique per page ([there is a 2019 feature request for having properties for more than one page](https://github.com/jgraph/drawio/issues/480)). So, because I already use Python to generate/adjust my drawio documents, I would like to generate these properties and copy them to each page inside a document. They look like this:

```
<root>
  <object label="" property1="abcdefg" property2="1.2.3.4" property3="micky mouse" id="pJwD9Pib3PC3m1MJIdun-0">
    <mxCell />
  </object>
```
...

However I'm trying to figure out how to generate the unique IDs. I've been searching through the code, but haven't found it yet. I was hoping someone can point me in the right direction.

**Comment by GuyDaniel** [comment]
I think it is like this:

```
public static final String TOKEN_ALPHABET = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ-_";
private static SecureRandom randomSecure = new SecureRandom();
	/**
	 * Returns a random string of the given length.
	 */
	public static String generateToken(int length)
	{
		StringBuffer rtn = new StringBuffer();

		for (int i = 0; i < length; i++)
		{
			int offset = randomSecure.nextInt(TOKEN_ALPHABET.length());
			rtn.append(TOKEN_ALPHABET.substring(offset,offset+1));
		}

		return rtn.toString();
	};
```
